modules/network_synchronizer/scene_synchronizer.cpp

```diff
@@ -0,0 +1,1845 @@
+	const int lowest_priority_number = INT32_MAX;
+	set_process_priority(lowest_priority_number);
+	set_physics_process_internal(true);
+
+	if (synchronizer) {
+		// Notify the presence all available nodes and its variables to the synchronizer.
+		for (uint32_t i = 0; i < node_data.size(); i += 1) {
+			synchronizer->on_node_added(node_data[i]);
+			for (int y = 0; y < node_data[i]->vars.size(); y += 1) {
+				synchronizer->on_variable_added(node_data[i], node_data[i]->vars[y].var.name);
+			}
+		}
+	}
+}
+
+void SceneSynchronizer::clear() {
+	if (synchronizer_type == SYNCHRONIZER_TYPE_NONETWORK) {
+		__clear();
+	} else {
+		ERR_FAIL_COND_MSG(get_tree()->is_network_server() == false, "The clear function must be called on server");
+		__clear();
+		rpc("__clear");
+	}
+}
+
+void SceneSynchronizer::__clear() {
+	for (uint32_t i = 0; i < node_data.size(); i += 1) {
+		Node *node = static_cast<Node *>(ObjectDB::get_instance(node_data[i]->instance_id));
+		if (node != nullptr) {
+			for (int y = 0; y < node_data[i]->vars.size(); y += 1) {
+				// Unregister the variable so the connected variables are
+				// correctly removed
+				unregister_variable(node, node_data[i]->vars[y].var.name);
+			}
+		}
+	}
+
+	node_data.clear();
+	controllers_node_data.clear();
+	global_nodes_node_data.clear();
+	node_counter = 1;
+
+	if (synchronizer) {
+		synchronizer->clear();
+	}
+}
+
+void SceneSynchronizer::_rpc_send_state(Variant p_snapshot) {
+	ERR_FAIL_COND(get_tree()->is_network_server() == true);
+	synchronizer->receive_snapshot(p_snapshot);
+}
+
+void SceneSynchronizer::_rpc_notify_need_full_snapshot() {
+	ERR_FAIL_COND(get_tree()->is_network_server() == false);
+
+	const int sender_peer = get_tree()->get_multiplayer()->get_rpc_sender_id();
+	PeerData *pd = peer_data.lookup_ptr(sender_peer);
+	ERR_FAIL_COND(pd == nullptr);
+	pd->need_full_snapshot = true;
+}
+
+void SceneSynchronizer::update_peers() {
+	if (peer_dirty == false) {
+		return;
+	}
+	peer_dirty = false;
+
+	for (uint32_t i = 0; i < controllers_node_data.size(); i += 1) {
+		PeerData *pd = peer_data.lookup_ptr(controllers_node_data[i]->node->get_network_master());
+		if (pd) {
+			pd->controller_id = controllers_node_data[i]->instance_id;
+		}
+	}
+}
+
+SceneSynchronizer::NodeData *SceneSynchronizer::register_node(Node *p_node) {
+	ERR_FAIL_COND_V(p_node == nullptr, nullptr);
+
+	NodeData *nd = get_node_data(p_node->get_instance_id());
+	if (unlikely(nd == nullptr)) {
+		const uint32_t node_id(generate_id ? node_counter++ : 0);
+		nd = memnew(NodeData);
+		nd->id = node_id;
+		nd->instance_id = p_node->get_instance_id();
+		nd->node = p_node;
+		node_data.push_back(nd);
+
+		NetworkedController *controller = Object::cast_to<NetworkedController>(p_node);
+		if (controller) {
+			if (unlikely(controller->has_scene_synchronizer())) {
+				node_data.erase(nd);
+				memdelete(nd);
+				ERR_FAIL_V_MSG(nullptr, "This controller has already a synchronizer. This is a bug!");
```

Comment:
Nit-picky: "already has" is more correct.

Reply:
Done ty